File: typescript/computer.ts
""
import InputDevice, {Mouse, Keyboard} from "./inputDevices";
import StorageDevice, {SSD, HDD} from "./storageDevices";
import ProcessingDevice, {CPU, GPU} from "./processingDevices";
import OutputDevice, {Monitor, Printer} from "./outputDevices";



// WiFi & Bluetooth Devices

interface WiFi {
    connectToWifi();
    disconnectFromWiFi();
}

interface Bluetooth {
    connectToBluetooth()
    disconnectFromBluetooth()
}


// The Computer Interface
class Computer {
    private _inputDevice: InputDevice;
    private _storageDevice: StorageDevice;
    private _processingDevice: ProcessingDevice;
    private _outputDevice: OutputDevice;

    constructor(
        inputDevice: InputDevice,
        storageDevice: StorageDevice,
        processingDevice: ProcessingDevice,
        outputDevice: OutputDevice,
        wifi?: WiFi,
        bluetooth?: Bluetooth
    ) {
        this._inputDevice = inputDevice;
        this._storageDevice = storageDevice;
        this._processingDevice = processingDevice;
        this._outputDevice = outputDevice;
    }

    setInputDevice(device: InputDevice): void {
        this._inputDevice = device;
    }

    setStorageDevice(device: StorageDevice): void {
        this._storageDevice = device;
    }

    setProcessingDevice(device: ProcessingDevice): void {
        this._processingDevice = device;
    }

    setOutputDevice(device: OutputDevice): void {
        this._outputDevice = device;
    }

    performInput(): string {
        return this._inputDevice.input();
    }

    performStorage(): string {
        return this._storageDevice.store() + " and " + this._storageDevice.retrieve();
    }

    performProcessing(): string {
        return this._processingDevice.process();
    }

    performOutput(): string {
        return this._outputDevice.output();
    }

}


class Laptop extends Computer implements WiFi, Bluetooth{

    connectToWifi(): string {
        return "WiFi connected";
    }
    disconnectFromWiFi(): string {
        return "WiFi disconnected";
    }
 
    connectToBluetooth(): string {
        return "Bluetooth connected";
    }
    disconnectFromBluetooth(): string {
        return "Bluetooth disconnected";
    }

}



// Example usage
let keyboard = new Keyboard();
let mouse = new Mouse();
let ssd = new SSD();
let cpu = new CPU();
let monitor = new Monitor();


// Desktop does not support WiFi and Bluetooth technologies
let computer = new Computer(keyboard, ssd, cpu, monitor);
console.log("========= COMPUTER =================================");
console.log(computer.performInput());      // Outputs: Keyboard input
console.log(computer.performStorage());    // Outputs: SSD storing data and SSD retrieving data
console.log(computer.performProcessing()); // Outputs: CPU processing data
console.log(computer.performOutput());     // Outputs: Monitor output
computer.setInputDevice(mouse);
console.log(computer.performInput());      // Outputs: Mouse input



// Laptop supports WiFi and Bluetooth technologies
let laptop = new Laptop(keyboard, ssd, cpu, monitor);
console.log("========= LAPTOP =================================");
console.log(laptop.performInput());      // Outputs: Keyboard input
console.log(laptop.performStorage());    // Outputs: SSD storing data and SSD retrieving data
console.log(laptop.performProcessing()); // Outputs: CPU processing data
console.log(laptop.performOutput());     // Outputs: Monitor output
laptop.setInputDevice(mouse);
console.log(laptop.performInput());      // Outputs: Mouse input
console.log(laptop.connectToWifi());       // Outputs: WiFi connected
console.log(laptop.connectToBluetooth());  // Outputs: Bluetooth connected

File: python/computer.py
from abc import ABC, abstractmethod
from input_device import InputDevice, Keyboard, Mouse
from storage_device import StorageDevice, SSD, HDD
from processing_device import ProcessingDevice, CPU, GPU
from output_device import OutputDevice, Monitor, Printer
from wifi_device import WiFiDevices
from bluetooth_device import BluetoothDevices


# Define Computer class
class Computer:
    def __init__(
        self,
        input_device: InputDevice,
        storage_device: StorageDevice,
        processing_device: ProcessingDevice,
        output_device: OutputDevice,
    ):
        self._input_device = input_device
        self._storage_device = storage_device
        self._processing_device = processing_device
        self._output_device = output_device

    def set_input(self, device: InputDevice) -> None:
        self._input_device = device

    def set_storage(self, device: StorageDevice) -> None:
        self._storage_device = device

    def set_processing(self, device: ProcessingDevice) -> None:
        self._processing_device = device

    def set_output(self, device: OutputDevice) -> None:
        self._output_device = device

    def input(self) -> str:
        return self._input_device.input()

    def storage(self) -> str:
        return f"{self._storage_device.store()} and {self._storage_device.retrieve()}"

    def processing(self) -> str:
        return self._processing_device.process()

    def output(self) -> str:
        return self._output_device.output()


# Laptop extends Computer and implements wifi and Bluetooth interfaces
class Laptop(Computer, WiFiDevices, BluetoothDevices):

    def connectToBluetooth(self) -> str:
        return "Bluetooth connected"

    def disconnectFromBluetooth(self) -> str:
        return "Bluetooth disconnected"

    def connectToWifi(self) -> str:
        return "WiFi connected"

    def disconnectFromWifi(self) -> str:
        return "WiFi disconnected"


# Example usage
keyboard = Keyboard()
mouse = Mouse()
ssd = SSD()
cpu = CPU()
monitor = Monitor()


computer = Computer(keyboard, ssd, cpu, monitor)
print("=================Computer===============")
print(computer.input())  # Outputs: Keyboard input
print(computer.storage())  # Outputs: SSD storing data and SSD retrieving data
print(computer.processing())  # Outputs: CPU processing data
print(computer.output())  # Outputs: Monitor output
# Change the input device
computer.set_input(mouse)
print(computer.input())  # Outputs: Mouse input
# Create a computer without WiFi and Bluetooth support


Laptop1 = Laptop(keyboard, ssd, cpu, monitor)

print("=================Laptop===============")
print(Laptop1.connectToBluetooth())  # Outputs: Bluetooth connected
print(Laptop1.disconnectFromBluetooth())  # Outputs: Bluetooth disconnected
print(Laptop1.connectToWifi())  # Outputs: WiFi connected
print(Laptop1.disconnectFromWifi())  # Outputs:WiFi disconnected
